File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/AppIntroArtworkSupport.kt
package com.pillyliu.pinprofandroid

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.dp
import com.pillyliu.pinprofandroid.ui.PinballThemeTokens

@Composable
internal fun AppIntroArtworkFrame(
    card: AppIntroCard,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .fillMaxWidth()
            .aspectRatio(card.artworkAspectRatio),
    ) {
        AppIntroArtworkBox(
            card = card,
            modifier = Modifier.matchParentSize(),
        )
    }
}

@Composable
internal fun AppIntroArtworkBox(
    card: AppIntroCard,
    modifier: Modifier = Modifier,
) {
    val accent = card.resolvedAccent()
    val shape = RoundedCornerShape(24.dp)

    Box(
        modifier = modifier
            .shadow(
                elevation = 16.dp,
                shape = shape,
                ambientColor = AppIntroTheme.tint.copy(alpha = 0.22f),
                spotColor = AppIntroTheme.tint.copy(alpha = 0.22f),
            )
            .clip(shape)
            .background(PinballThemeTokens.colors.atmosphereBottom.copy(alpha = 0.99f))
            .background(
                Brush.linearGradient(
                    colors = listOf(
                        AppIntroTheme.tint.copy(alpha = 0.26f),
                        Color.Black.copy(alpha = 0.14f),
                        PinballThemeTokens.colors.brandGold.copy(alpha = 0.10f),
                    ),
                ),
            )
            .border(1.15.dp, accent.copy(alpha = 0.72f), shape),
    ) {
        if (card == AppIntroCard.Welcome) {
            Image(
                painter = painterResource(id = card.artworkResId),
                contentDescription = null,
                modifier = Modifier.fillMaxSize(),
                contentScale = ContentScale.Crop,
            )
        } else {
            Box(
                modifier = Modifier.fillMaxSize(),
                contentAlignment = Alignment.Center,
            ) {
                Box(
                    modifier = Modifier
                        .matchParentSize()
                        .background(
                            Brush.radialGradient(
                                colors = listOf(
                                    accent.copy(alpha = 0.18f),
                                    Color.Transparent,
                                ),
                                radius = 480f,
                            ),
                        ),
                )
                Image(
                    painter = painterResource(id = card.artworkResId),
                    contentDescription = null,
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(1.dp)
                        .shadow(
                            elevation = 16.dp,
                            spotColor = accent.copy(alpha = 0.16f),
                            ambientColor = accent.copy(alpha = 0.16f),
                        ),
                    contentScale = ContentScale.Fit,
                )
            }
        }
    }
}

@Composable
internal fun AppIntroProfessorSpotlight(
    side: AppIntroProfessorSide,
) {
    Box(
        modifier = Modifier.size(82.dp),
        contentAlignment = Alignment.Center,
    ) {
        Box(
            modifier = Modifier
                .matchParentSize()
                .background(
                    Brush.radialGradient(
                        colors = listOf(
                            AppIntroTheme.glow.copy(alpha = 0.34f),
                            AppIntroTheme.tint.copy(alpha = 0.20f),
                            Color.Transparent,
                        ),
                        radius = 120f,
                    ),
                    CircleShape,
                ),
        )
        Box(
            modifier = Modifier
                .size(72.dp)
                .background(Color.White.copy(alpha = 0.04f), CircleShape)
                .border(1.dp, Color.White.copy(alpha = 0.10f), CircleShape),
        )
        Image(
            painter = painterResource(id = R.drawable.intro_professor_headshot),
            contentDescription = null,
            modifier = Modifier
                .size(80.dp)
                .offset(y = (-2).dp)
                .clip(CircleShape)
                .graphicsLayer {
                    scaleX = if (side == AppIntroProfessorSide.Left) -1f else 1f
                },
            contentScale = ContentScale.Crop,
        )
    }
}

File: Pinball App 2/Pinball App 2/app/AppIntroArtworkSupport.swift
import SwiftUI
import UIKit

struct AppIntroArtworkFrame: View {
    let card: AppIntroCard

    var body: some View {
        Color.clear
            .aspectRatio(card.artworkAspectRatio, contentMode: .fit)
            .overlay {
                AppIntroArtworkBox(card: card)
            }
    }
}

struct AppIntroArtworkBox: View {
    let card: AppIntroCard

    var body: some View {
        let shape = RoundedRectangle(cornerRadius: 24, style: .continuous)

        ZStack {
            shape
                .fill(AppTheme.atmosphereBottom.opacity(0.99))
                .overlay(
                    shape
                        .fill(
                            LinearGradient(
                                colors: [
                                    AppIntroTheme.tint.opacity(0.26),
                                    Color.black.opacity(0.14),
                                    AppTheme.brandGold.opacity(0.10)
                                ],
                                startPoint: .topLeading,
                                endPoint: .bottomTrailing
                            )
                        )
                )

            Group {
                if card == .welcome {
                    AppIntroWelcomeArtwork()
                } else {
                    AppIntroScreenshotArtwork(
                        bundledFileName: card.bundledArtworkFileName,
                        accent: card.accent
                    )
                }
            }
        }
        .clipShape(shape)
        .overlay(
            shape
                .stroke(card.accent.opacity(0.72), lineWidth: 1.15)
        )
        .shadow(color: AppIntroTheme.tint.opacity(0.22), radius: 16, y: 8)
    }
}

private struct AppIntroWelcomeArtwork: View {
    var body: some View {
        Group {
            if let image = AppIntroBundledArtProvider.image(named: AppIntroCard.welcome.bundledArtworkFileName) {
                Image(uiImage: image)
                    .resizable()
                    .scaledToFill()
            } else {
                Color.clear
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .clipped()
    }
}

private struct AppIntroScreenshotArtwork: View {
    let bundledFileName: String
    let accent: Color

    var body: some View {
        ZStack {
            RadialGradient(
                colors: [accent.opacity(0.18), .clear],
                center: .center,
                startRadius: 14,
                endRadius: 240
            )

            if let image = AppIntroBundledArtProvider.image(named: bundledFileName) {
                Image(uiImage: image)
                    .resizable()
                    .scaledToFit()
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                    .shadow(color: accent.opacity(0.16), radius: 16, y: 8)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

struct AppIntroProfessorSpotlight: View {
    let side: AppIntroProfessorSide

    var body: some View {
        ZStack {
            Circle()
                .fill(
                    RadialGradient(
                        colors: [
                            AppIntroTheme.glow.opacity(0.34),
                            AppIntroTheme.tint.opacity(0.20),
                            .clear
                        ],
                        center: .center,
                        startRadius: 8,
                        endRadius: 54
                    )
                )
                .frame(width: 82, height: 82)

            Circle()
                .fill(Color.white.opacity(0.04))
                .frame(width: 72, height: 72)
                .overlay(
                    Circle()
                        .stroke(Color.white.opacity(0.10), lineWidth: 1)
                )
                .overlay {
                    if let image = AppIntroBundledArtProvider.image(named: "professor-headshot.webp") {
                        Image(uiImage: image)
                            .resizable()
                            .scaledToFill()
                            .frame(width: 80, height: 80)
                            .scaleEffect(x: side == .left ? -1 : 1, y: 1)
                            .offset(y: -2)
                            .clipShape(Circle())
                    }
                }
        }
        .frame(width: 82, height: 82)
    }
}

enum AppIntroBundledArtProvider {
    private static let cache = NSCache<NSString, UIImage>()

    static func image(named fileName: String) -> UIImage? {
        let cacheKey = fileName as NSString
        if let cached = cache.object(forKey: cacheKey) {
            return cached
        }

        guard let url = bundledURL(named: fileName),
              let data = try? Data(contentsOf: url),
              let image = UIImage(data: data) else {
            NSLog("Missing bundled intro artwork: %@", fileName)
            return nil
        }

        cache.setObject(image, forKey: cacheKey)
        return image
    }

    private static func bundledURL(named fileName: String) -> URL? {
        if let rootURL = Bundle.main.url(forResource: fileName, withExtension: nil) {
            return rootURL
        }
        return Bundle.main.url(
            forResource: fileName,
            withExtension: nil,
            subdirectory: "SharedAppSupport/app-intro"
        )
    }
}
